### PR TITLE
feat: ULID

### PR DESCRIPTION
**Pull request by Cobollatin**
Hey, it's my first time contributing to C++, so I may require some patience. I just added this implementation: https://github.com/suyash/ulid to the lib for ULID(https://github.com/cieslarmichal/faker-cxx/issues/997). I created a test and everything is building/passing. Just a note, when I run the format script, I get `Invalid value for—style` so it may be a version issue. Which version of clang-format should I use?

### CODE REVIEW

File: tests/modules/string_test.cpp [hunk 1]
```diff
@@ -124,6 +124,13 @@ TEST_F(StringTest, shouldGenerateUuid4)
     ASSERT_EQ(generatedUuid[23], '-');
 }
 
+TEST_F(StringTest, shouldGenerateUlid)
```
Comment (cieslarmichal) [inline tests/modules/string_test.cpp:127]: add also a test without argument

File: include/faker-cxx/string.h [hunk 2]
```diff
@@ -118,6 +120,34 @@ std::string uuid(RandomGenerator<T> gen = RandomGenerator<std::mt19937>{})
...
+
+ // Based on https://github.com/suyash/ulid
+
+template <typename T = time_t>
+std::string ulid(time_t refDate = std::time(nullptr))
+{
+    const auto uild = ulid_ext::Create(refDate, []() { return 4; });
+    std::string data = ulid_ext::Marshal(uild);
+    return std::string(data);
+}
```
Comment (cieslarmichal) [inline include/faker-cxx/string.h:149]: could we move implementation to string.cpp?

File: include/helpers/ulid/ulid_struct.hh [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+#include <random>
+#include <vector>
+
+#if _MSC_VER > 0
+typedef uint32_t rand_t;
+# else
+typedef uint8_t rand_t;
+#endif
+
+namespace ulid_ext {
```
Comment (cieslarmichal) [inline include/helpers/ulid/ulid_struct.hh:17]: it should be inside faker::helpers namespace

File: include/helpers/ulid/ulid.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+// http://stackoverflow.com/a/23981011
+#ifdef __SIZEOF_INT128__
+#define ULIDUINT128
+#endif
+
+#ifdef ULIDUINT128
+#include "ulid_uint128.hh"
+#else
+#include "ulid_struct.hh"
```
Comment (cieslarmichal) [inline include/helpers/ulid/ulid.hpp:12]: change headers extensions from `.hpp`, `.hh` to just `.h`

File: include/helpers/ulid/ulid_struct.hh [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+
+		data[14] = other.data[14];
+		other.data[14] = 0;
+
+		data[15] = other.data[15];
+		other.data[15] = 0;
+	}
+
+	ULID& operator=(ULID&& other) {
+		// for (int i = 0 ; i < 16 ; i++) {
```
Comment (cieslarmichal) [inline include/helpers/ulid/ulid_struct.hh:196]: please remove comments

File: include/helpers/ulid/ulid_struct.hh [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+
+/**
+ * CompareULIDs will compare two ULIDs.
+ * returns:
+ *     -1 if ulid1 is Lexicographically before ulid2
+ *      1 if ulid1 is Lexicographically after ulid2
+ *      0 if ulid1 is same as ulid2
+ * */
+inline int CompareULIDs(const ULID& ulid1, const ULID& ulid2) {
+	// for (int i = 0 ; i < 16 ; i++) {
```
Comment (cieslarmichal) [inline include/helpers/ulid/ulid_struct.hh:605]: remove comments

File: include/helpers/ulid/ulid_uint128.hh [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+#include <random>
+#include <vector>
+
+#if _MSC_VER > 0
+typedef uint32_t rand_t;
+# else
+typedef uint8_t rand_t;
+#endif
+
+namespace ulid_ext {
```
Comment (cieslarmichal) [inline include/helpers/ulid/ulid_uint128.hh:17]: faker::helpers namespace

**Review by cieslarmichal — CHANGES_REQUESTED** [review]
Please also add a section in LICENSES file and mention this repo you used

**Comment by cieslarmichal** [comment]
please check if we can remove some methods from this implementation, maybe we don't use every function from his repo?

**Comment by Cobollatin** [comment]
> Please also add a section in LICENSES file and mention this repo you used

Done

> please check if we can remove some methods from this implementation, maybe we don't use every function from his repo?

Yes, I will keep only the minimum.

**Review by cieslarmichal — APPROVED** [review]

**Comment by cieslarmichal** [comment]
I have clang-format in version 18. please try to install newer than you have

**Comment by cieslarmichal** [comment]
and run formatting

**Comment by Cobollatin** [comment]
> and run formatting
I get the same error
`Invalid value for -style`
I am using version 18, any idea why?
```
$clang-format --version
Ubuntu clang-format version 18.1.8 (++20240731024944+3b5b5c1ec4a3-1~exp1~20240731145000.144)
```

**Comment by cieslarmichal** [comment]
Build also fails

**Comment by Cobollatin** [comment]
> Build also fails

Last commit should fix shared-lib builds.

**Comment by Cobollatin** [comment]
I ran the format command manually and it worked, seems like the script has an issue

**Comment by Cobollatin** [comment]
image
It works when I add the bash path, should I also commit this file?

**Comment by cieslarmichal** [comment]
Yes please commit

**Comment by Cobollatin** [comment]
> Yes please commit

Done

**Pull request merged**